sdk/webpubsub/azure-messaging-webpubsub-client/src/main/java/com/azure/messaging/webpubsub/client/implementation/websocket/WebSocketClientHandler.java

```diff
@@ -46,80 +52,136 @@ ChannelFuture handshakeFuture() {
+
+        if (webSocketFrame instanceof PingWebSocketFrame) {
             // Ping, reply Pong
             loggerReference.get().atVerbose().log("Received PingWebSocketFrame");
             loggerReference.get().atVerbose().log("Send PongWebSocketFrame");
-            ch.writeAndFlush(new PongWebSocketFrame());
-        } else if (frame instanceof PongWebSocketFrame) {
+            channel.writeAndFlush(new PongWebSocketFrame());
+            return true;
+        } else if (webSocketFrame instanceof PongWebSocketFrame) {
             // Pong
             loggerReference.get().atVerbose().log("Received PongWebSocketFrame");
-        } else if (frame instanceof CloseWebSocketFrame) {
+            return true;
+        } else if (webSocketFrame instanceof CloseWebSocketFrame) {
             // Close
-            CloseWebSocketFrame closeFrame = (CloseWebSocketFrame) frame;
+            final CloseWebSocketFrame closeFrame = (CloseWebSocketFrame) webSocketFrame;
+
             loggerReference.get()
                 .atVerbose()
                 .addKeyValue("statusCode", closeFrame.statusCode())
                 .addKeyValue("reasonText", closeFrame.reasonText())
                 .log("Received CloseWebSocketFrame");
-
-            this.serverCloseWebSocketFrame = closeFrame.retain();   // retain for SessionNettyImpl
+            serverCloseWebSocketFrame = closeFrame.retain();   // retain for SessionNettyImpl
 
             if (closeCallbackFuture == null) {
                 // close initiated from server, reply CloseWebSocketFrame, then close connection
                 loggerReference.get().atVerbose().log("Send CloseWebSocketFrame");
                 closeFrame.retain();    // retain before write it back
-                ch.writeAndFlush(closeFrame).addListener(future -> ch.close());
+                channel.writeAndFlush(closeFrame).addListener(future -> channel.close());
             } else {
                 // close initiated from client, client already sent CloseWebSocketFrame
-                ch.close();
+                channel.close();
             }
+
+            return true;
+        } else if (webSocketFrame instanceof TextWebSocketFrame
+            || webSocketFrame instanceof ContinuationWebSocketFrame) {
+            if (compositeByteBuf == null) {
+                compositeByteBuf = ByteBufAllocator.DEFAULT.compositeBuffer();
+            }
+
+            compositeByteBuf.addComponent(true, webSocketFrame.content().retain());
+
+            if (!webSocketFrame.isFinalFragment()) {
+                return true;
+            }
+
+            publishBuffer();
+            return true;
+        } else {
+            return false;
         }
     }
 
-    @Override
-    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
-        cause.printStackTrace();
-        if (handshakeFuture != null && !handshakeFuture.isDone()) {
-            handshakeFuture.setFailure(cause);
+    private void publishBuffer() {
```

Comment:
nit; I wonder if `handlerRemoved` could use this function (looks like same code)?

Reply:
Yup! I had consolidated it into a single function for that reason. :D